### ISSUE TITLE
Intelligent planned dispatch slot not turning on

### ISSUE THREAD

**Issue by vansteedens** [megakid/ha_octopus_intelligent#35]
I utilise the planned dispatch slot to trigger automations and stop my house batteries from draining. Recently, I've noticed that the slot is never turning on, instead it goes from off to unknown and then unknown to off. This means that my battery automations aren't correctly firing and updating the discharge settings. I can't use the 'any state' option either because unknown doesn't necessarily mean it's in a dispatch slot. The only time any state is useful is when turning off the automation, however this then fires at random times in the day when the service becomes unavailable. You can see the last few days examples in the screenshot. The car was charging between 23:00 - 07:30 with a few 30 slots of stopping.

![Screenshot 2024-02-16 at 09 31 26](https://github.com/megakid/ha_octopus_intelligent/assets/147421043/e57ea36f-679a-470b-b804-406b6dd9e913)

**Comment by pdcastro** [comment]
Are you running version 1.6.3 [released](https://github.com/megakid/ha_octopus_intelligent/releases) two weeks ago? Since I updated to that release, stability improved significantly. I still see occasional ~5min unavailable gaps as in your screenshot, but I see ‘turned on’ slots too:

![slots](https://github.com/megakid/ha_octopus_intelligent/assets/15091591/9766bc17-1a74-4774-8917-c88abe1b70c2)

I am currently running the latest HASS version `2024.2.1`, and for the past week or so I was running HASS version `2024.1.6`.

**Comment by vansteedens** [comment]
I'm running the same versions as you so not sure what's going on and why it's not changed to on for a while. I've changed my automations over to the Bottlecapdave version for now as the dispatching sensor on there seems to have recorded all my recent sessions.

**Comment by matt7aylor** [comment]
Hi, I have been having the same issue (running latest v1.6.3). After a little digging I believe have found what was causing the problem for me and a temporary workaround. (I need the workaround as I rely on the planned dispatch slot to automate turning on the car charger to actually charge the car.) However, I think the problem might lie in the Octopus API, unfortunately (if we can confirm this does anyone know how we can report it?).

It seem that the `plannedDispatches` metadata returned by the API changes after the schedule is updated (by Octopus). Once the charge schedule first appears for each entry in `plannedDispatches` the metadata will look like `{'meta': {'source': 'smart-charge', ...}}`, which is the expected behaviour and the planned dispatch slot entity will work if the schedule is not changed before it begins (sometimes I would get working dispatches in the morning after fiddling with the settings). However, it seems that the schedule often gets altered after it is set the first time and then the metadata seems to change instead to `{'meta': {'source': None, ...}}`. This causes the planned dispatch slot to not register in the integration.

A workaround that allows me to still charge the car in the dispatch slot is to change the line 
https://github.com/megakid/ha_octopus_intelligent/blob/9f45e5987abdee6ddf6d1c190a61f14b7e20e513/custom_components/octopus_intelligent/octopus_intelligent_system.py#L126

to:
```python
            if source is None or state.get('meta', {}).get('source', '') in [source, None]:  # HACK workaround empty meta
```

This works for me since I don't use the "bump charge" feature but this may not be a good permanent solution as it could potentially result in peak time bump charges being shown in the planned dispatch slot, which would be a change from the current behaviour as I understand it.

**Comment by stintov** [comment]
I have the same issue. I’m on 1.6.3 but I’ve been having the issue for two weeks now.

**Comment by pdcastro** [comment]
> _”However, it seems that the schedule often gets altered after it is set the first time and then the metadata seems to change instead to {'meta': {'source': None, ...}}.”_

Great findings @matt7aylor! 👍 Given your explanation that the metadata `source` is initially valid (`'smart-charge'` or `'bump-charge'`) but then becomes `None`, how about saving the “last seen” valid `source` value in a variable, and assuming the saved value whenever source is `None` in the metadata? I assume that planned dispatches cannot simultaneously be ‘smart’ and ‘bump’, so the last seen source (other than `None`) is the correct one.

**Comment by matt7aylor** [comment]
The problem I think is that the metadata is lost when the `plannedDispatches` changes so where before you might have had a dispatch e.g. from 01:00 to 01:30 with `smart-charge`, now you might have one 04:30 to 05:00 with `None`. I'm not sure how reliable it will be to remember them and try to guess metadata for the new ones, because maybe the new one is not actually a smart-charge.

**Comment by stintov** [comment]
Any update on this please? I’ve found that running the automation manually works. 

I believe that as only started after applying the last 1.6.3 integration.

**Comment by matt7aylor** [comment]
@stintov Have you tried the workaround here https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197 , works perfectly for me.

**Comment by stintov** [comment]
> @stintov Have you tried the workaround here [#35 (comment)](https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197) , works perfectly for me.

Thanks @matt7aylor, I will put my big boys pants on and try and alter the code. I can't say I know what I'm doing here :). Was hoping the dev would write a fix for us.

**Comment by stintov** [comment]
> @stintov Have you tried the workaround here [#35 (comment)](https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197) , works perfectly for me.

![image](https://github.com/megakid/ha_octopus_intelligent/assets/130797662/5c17ad1b-4142-495b-aa1b-fb4239021fd1)

Lets see if this works now. Great work if this fixes it.

**Comment by davebarrett1959** [comment]
Was having the same issue but using the "Octopus Intelligent Slot" in my HA NodeRed automations. [https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197](https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197) has fixed it. Many thanks @matt7aylor

**Comment by pdcastro** [comment]
> _[...] How about saving the “last seen” valid `source` value in a variable, and assuming the saved value whenever `source` is None in the metadata?_
>> _[...] I'm not sure how reliable it will be to remember them and try to guess metadata for the new ones, because maybe the new one is not actually a smart-charge._

With two weeks of testing, I have found the “last seen valid source” approach to be effective in practice and I submitted PR #37 for it. Empirically, so far it has held true that when the source changes from `smart-charge` to `bump-charge` or vice-versa, then at least one non-None value appears in the metadata with the integration's 5-minute polling interval. If Home Assistant is down for maintenance for a longer period of time ***and*** bump charging is started or stopped before the integration gets to poll the API again, then there is a _chance_ that this solution would mistake smart charging for bump charging, but even then it would fix itself as soon as the Octopus API returned the first non-None `source` value. And if one only ever cared about smart charging, then the “last seen valid source” solution _always_ gets it right, so it is at least as effective as the hot patch in https://github.com/megakid/ha_octopus_intelligent/issues/35#issuecomment-1949942197.

**Comment by megakid** [comment]
Hey @pdcastro thanks so much for all the activity here. Would you like to become a maintainer so you can do merges? I'm not able to dedicate the time I'd like to this

**Comment by pdcastro** [comment]
@megakid, thanks for the invite. I wrote to you at the email address in your public GitHub profile. Could my emails be ending up in your spam folder? Even after accepting your GitHub invite, merging the PRs still requires the approval of a reviewer with write permissions, _other than the PR author._

**Comment by megakid** [comment]
Thanks @pdcastro I have approved - hopefully I can alter the settings to allow you to unilaterally merge PRs

**Comment by pdcastro** [comment]
GitHub automatically closed this issue because of the ‘fix’ verb used in the message of a commit that was merged to the main branch, however a tagged release hasn’t been created yet. Reopening this issue until the release is created.

**Comment by hammerhead2000** [comment]
Personally I think 1.7.0 is fine, I think users expect significant new features if you change from 1.x to 2.x.

Thank you for fixing this, I was about to start digging through the code myself as this problem is breaking my AppDaemon automations!  I would happily pull this as a beta release to give you some extra testing if that would be helpful.

**Comment by pdcastro** [comment]
Thanks @hammerhead2000 and @stintov 👍  I went ahead and published release [v1.7.0](https://github.com/megakid/ha_octopus_intelligent/releases/tag/v1.7.0). It’s already available for installation through HACS (custom repositories) as well. Hopefully it [sufficiently](https://github.com/megakid/ha_octopus_intelligent/pull/37#issuecomment-1979168329) solves this issue. 🤞

**Comment by pdcastro** [comment]
> _[...] changed my automations over to the Bottlecapdave version for now as the dispatching sensor on there seems to have recorded all my recent sessions._

On this note, I’ve been running Bottlecapdave’s integration with debug logging out of interest, as I wanted to see if somehow it was immune to this issue that we believe to be an issue with Octopus’ API. Today I saw the issue happening to me with Bottlecapdave’s integration:

```
2024-03-20 00:01:05.606 DEBUG (MainThread) [custom_components.octopus_energy.api_client] async_get_intelligent_dispatches: {'data': {
    'plannedDispatches': [{
        'startDt': '2024-03-20 00:00:00+00:00',
        ...
        'meta': {'source': None, 'location': None}
    }], ...
```

So there we go: This issue affects Bottlecapdave’s integration as well, and right now megakid’s integration is thus more reliable. (Well, apart from #43, but #44 is already awaiting to be released as 1.7.1...) 😄

**Comment by vansteedens** [comment]
> > _[...] changed my automations over to the Bottlecapdave version for now as the dispatching sensor on there seems to have recorded all my recent sessions._
> 
> On this note, I’ve been running Bottlecapdave’s integration with debug logging out of interest, as I wanted to see if somehow it was immune to this issue that we believe to be an issue with Octopus’ API. Today I saw the issue happening to me with Bottlecapdave’s integration:
> 
> ```
> 2024-03-20 00:01:05.606 DEBUG (MainThread) [custom_components.octopus_energy.api_client] async_get_intelligent_dispatches: {'data': {
>     'plannedDispatches': [{
>         'startDt': '2024-03-20 00:00:00+00:00',
>         ...
>         'meta': {'source': None, 'location': None}
>     }], ...
> ```
> 
> So there we go: This issue affects Bottlecapdave’s integration as well, and right now megakid’s integration is thus more reliable. (Well, apart from #43, but #44 is already awaiting to be released as 1.7.1...) 😄

Awesome, I switched back once you said it was fixed and going to do another charge tonight so hopefully all my automations work as expected 😁
